core(graph): document and exemplify capture (#652)
kokkos/kokkos-core-wiki: docs/source/ProgrammingGuide/Graph.rst

Such a scenario can be encountered in many situations like building and training a neural network,
running a conjugate gradient method, and so on.

Capturing into a :cppkokkos:`Kokkos::Graph` boils down to writing the following snippet:

.. code:: c++

    auto my_captured_node = predecessor.cuda_capture(
        exec,
        [data = data](const Kokkos::Cuda& exec_) { ... }
    );

When the node is added to the :cppkokkos:`Kokkos::Graph`, the workloads are not dispatched to the device.
Rather, the backend operations are "saved" for later "reuse" in the *capture* node.

This is achieved by setting the underlying "stream" in "*capture* mode" before invoking the function object,
such that backend operations are saved to the *capture* node instead of being enqueued in the passed execution space instance.

Some important aspects of *capture* are worth pointing out:

1. The function object will be stored by the :cppkokkos:`Kokkos::Graph` instance,
   thereby ensuring that any data bound to the function object is guarenteed to
   stay alive until the graph is destroyed.
2. The parameter `exec` can be used to assign the captured workloads to the device associated with
   `exec`.

> 3. The *capture* mechanism will temporarily alter the `exec` state by setting the underlying "stream" in "*capture* mode", such that

Reviewer: What are the documentation references for the other backends?

Author: * `SYCL` https://github.com/intel/llvm/blob/ee5e1ca95c78576c1b6f12b1c2d461ef4b537a9b/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc?plain=1#L167-LL170 is a good starting point, and the link will be valid forever so I'm fine adding it.
* `HIP`: https://rocm.docs.amd.com/projects/HIP/en/latest/reference/hip_runtime_api/modules/graph_management.html#_CPPv421hipStreamBeginCapture11hipStream_t20hipStreamCaptureMode but the link seems pretty fragile.

Reviewer: Hmm I prefer having the links and updating them if someone complains later.

  the backend driver and/or compiler can leverage optimization opportunities.
* Launch overhead is reduced, benefitting DAGs consisting of small workloads.

Capture
~~~~~~~

It might be necessary to add nodes to a :cppkokkos:`Kokkos::Graph` whose workloads are not defined

> using any :cppkokkos:`Kokkos` code.

Author: Done :heavy_check_mark:

This is the case *e.g.* when calling external math libraries like `cuBLAS`.

Such a scenario can be encountered in many situations like building and training a neural network,

> running a conjugate gradient method, and so on.

Author: Not sure either but I4d keep it.

> running a conjugate gradient method, and so on.

Reviewer 2: It's nice to make a link with domains that are a bit different from our usual domains like solvers :)

Capturing into a :cppkokkos:`Kokkos::Graph` boils down to writing the following snippet:

.. code:: c++

    auto my_captured_node = predecessor.cuda_capture(
        exec,
        [data = data](const Kokkos::Cuda& exec_) { ... }
    );

When the node is added to the :cppkokkos:`Kokkos::Graph`, the workloads are not dispatched to the device.
Rather, the backend operations are "saved" for later "reuse" in the *capture* node.

This is achieved by setting the underlying "stream" in "*capture* mode" before invoking the function object,

> such that backend operations are saved to the *capture* node instead of being enqueued in the passed execution space instance.

Author: Yes, I've added a sentence.

Some important aspects of *capture* are worth pointing out:

1. The function object will be stored by the :cppkokkos:`Kokkos::Graph` instance,
   thereby ensuring that any data bound to the function object is guarenteed to
   stay alive until the graph is destroyed.
2. The parameter `exec` can be used to assign the captured workloads to the device associated with

>    `exec`.

Author: Done :heavy_check_mark: Better indeed !

> 3. The *capture* mechanism will temporarily alter the `exec` state by setting the underlying "stream" in "*capture* mode", such that

Author: :heavy_check_mark:

>    backend-specific restrictions may apply (see `the Cuda programming guide <https://docs.nvidia.com/cuda/cuda-c-programming-guide/#prohibited-and-unhandled-operations>`_

Author: OK :heavy_check_mark:

  the backend driver and/or compiler can leverage optimization opportunities.
* Launch overhead is reduced, benefitting DAGs consisting of small workloads.

Capture
~~~~~~~

It might be necessary to add nodes whose workloads are not

> defined using any :cppkokkos:`Kokkos` code to a :cppkokkos:`Kokkos::Graph`.

Author: It's been done in #666 (:imp:) indeed !

> defined using any :cppkokkos:`Kokkos` code to a :cppkokkos:`Kokkos::Graph`.

Author: Done :heavy_check_mark:

> It might be necessary to add nodes whose workloads are not
> defined using any :cpp:`Kokkos` code to a :cpp:`Kokkos::Graph`.
> This is the case *e.g.* when calling external math libraries like `cuBLAS`.

Reviewer 3: ```suggestion
Some use cases might require adding nodes to a :cpp:`Kokkos::Graph` with workloads that aren't
expressed in terms of Kokkos API but rather in native code, e.g., calling external math libraries like `cuBLAS`.
```
sounds a but smoother in my opinion.

Author: OK :heavy_check_mark:

Such a scenario can be encountered in many situations like building and training a neural network,
running a conjugate gradient method, and so on.

Capturing into a :cppkokkos:`Kokkos::Graph` boils down to writing the following snippet:

.. code:: c++

    auto my_captured_node = predecessor.cuda_capture(
        exec,

>         [data = data](const Kokkos::Cuda& exec_) { ... }

Author: Not sure what would be best to avoid the lambda capture and graph capture confusion. Avoid the lambda and explicitly use a functor instead ?

>         [data = data](const Kokkos::Cuda& exec_) { ... }

Reviewer 3: Why exactly do you require a stateful lambda? AFAICT that highly depends on the use case.

Rather, the backend operations are "saved" for later "reuse" in the *capture* node.

This is achieved by setting the underlying "stream" to "*capture* mode" before invoking the function object,
such that backend operations are saved to the *capture* node instead of being enqueued in the passed execution space instance.
The underlying "stream" is then restored to "no-*capture* mode".

Some important aspects of *capture* are worth pointing out:

1. The function object will be stored by the :cppkokkos:`Kokkos::Graph` instance,
   thereby ensuring that any data bound to the function object is guaranteed to
   stay alive until the graph is destroyed.
2. The execution space instance `exec` associates the captured workloads to a device.
3. The *capture* mechanism will temporarily alter the `exec`'s state by setting the underlying "stream" to "*capture* mode".
   While in "*capture* mode", backend-specific restrictions may apply (see `the Cuda programming guide <https://docs.nvidia.com/cuda/cuda-c-programming-guide/#prohibited-and-unhandled-operations>`_
   for instance).

.. warning::

    When a "stream" is used by many threads, capturing on one thread may affect other threads

>     (see `CUDA`'s :cpp:`cudaThreadExchangeStreamCaptureMode` for instance).

Reviewer: do we have a link for that?

Author: Again, the link to `cudaThreadExchangeStreamCaptureMode` is ugly and seems prone to changes:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85

Are you sure you want it in the documentation? In the end, it's just going to https://docs.nvidia.com/cuda/cuda-runtime-api/ and searching for `cudaThreadExchangeStreamCaptureMode`.

Author: Updated, OK for you ?

>     (search for :cpp:`cudaThreadExchangeStreamCaptureMode` on `Cuda runtime API docmentation <https://docs.nvidia.com/cuda/cuda-runtime-api/>`_ for instance).

Reviewer 4: ```suggestion
    (search for :cpp:`cudaThreadExchangeStreamCaptureMode` on `Cuda runtime API documentation <https://docs.nvidia.com/cuda/cuda-runtime-api/>`_ for instance).
```

Author: :heavy_check_mark:

For now, *capture* is only supported for the following backends:

.. list-table::

  * - Backend
    - Resources
  * - :cppkokkos:`Cuda`

>     - `CUDA Graphs <https://docs.nvidia.com/cuda/cuda-c-programming-guide/#cuda-graphs>`_

Reviewer 4: Should we be referring to https://docs.nvidia.com/cuda/cuda-c-programming-guide/#creating-a-graph-using-stream-capture and say "CUDA Stream Capture" instead of CUDA Graphs?

Author: Yep, better indeed :+1:

  * - :cppkokkos:`HIP`
    - `HIP stream capture <https://rocm.docs.amd.com/projects/HIP/en/latest/reference/hip_runtime_api/modules/graph_management.html#_CPPv421hipStreamBeginCapture11hipStream_t20hipStreamCaptureMode>`_
  * - :cppkokkos:`SYCL`

>     - `SYCL Graph <https://github.com/intel/llvm/blob/ee5e1ca95c78576c1b6f12b1c2d461ef4b537a9b/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc?plain=1#L167-LL170>`_

Reviewer 4: Should it be saying "SYCL queue recording"?

.. note::

>     The :cppkokkos:`SYCL` documentation will use *recording* instead of *capture*, but it is essentially the same thing.

Reviewer 4: ```suggestion
    The :cppkokkos:`SYCL` documentation will use the term *recording* instead of *capture*, but it is essentially the same thing.
```

Author: :heavy_check_mark:

    graph.instantiate();

    graph.submit();

Capture of a `cuBLAS` call
~~~~~~~~~~~~~~~~~~~~~~~~~~

This example shows how to create a node that captures a `cuBLAS` call.
It also demonstrates how data is kept alive by the :cppkokkos:`Kokkos::Graph`

> until it is destroyed (*e.g.* the `cuBLAS` handle).

Author: Reworded.

> until it is destroyed (*e.g.* the `cuBLAS` handle).

Author: Indeed, done :heavy_check_mark:
